### DISCUSSION TITLE
Using simdjson to check input is valid JSON without full parsing

### DISCUSSION THREAD

**Discussion by daverigby** [simdjson/simdjson#1393] [Q&A]
Hi all,

As per subject, I'm wondering what's the most efficient way to use simdjson to simply check an input is valid JSON, without actually parsing out the content. Essentially the same use-case as http://www.json.org/JSON_checker/

Ideally this would require state which _isn't_ proportional to the size of the input document; for my use-case I need to validate many MB JSON files without consuming a similar amount of memory.

Thanks in advance.

**Comment by daverigby** [comment]
Looking at the "classic" (not on-demand) API, it seems like you get this basically "automatically" when `parse()` is called, however that appears to use memory proportional to the JSON size which isn't suitable for my use-case.

I noticed the [Tree Walking](https://github.com/simdjson/simdjson/blob/master/doc/basics.md#tree-walking-and-json-element-types) example for the classic API; seems like something along those lines could work for the on-demand API (which doesn't use memory proportional to JSON size), but I can't see any functions in the on-demand API to return the type of what the iterator points to. 

I guess I _could_ forcefully try each possible JSON type, but that feels (a) at bit dirty and (b) maybe is slow? Is there a better approach I could use?

**Comment by lemire** [comment]
These are interesting questions but simdjson does not currently offer a validation-only function. That would be interesting, but there is just no such thing currently.

**Comment by daverigby** [comment]
Thanks for responding Daniel. 

For some more context, my use-case is ingesting JSON documents into a database system, where the documents can be anything the user wants - hence a schema-style validation as proposed by the on-demand API docs isn't applicable.

I'm investigating potential performance improvements to the JSON validation we currently have, and simdjson seemed an interesting candidate given the performance numbers quoted.

**Comment by lemire** [comment]
> schema-style validation as proposed by the on-demand API docs isn't applicable.

@jkeiser  What are your thoughts on this?

**Comment by jkeiser** [comment]
It's accurate ... the fastest way to fully validate is probably to run it through simdjson's Dom api.

It's possible to do with On Demand using recursion and a little elbow grease right now; I really don't know whether that will be faster or slower. In future releases on demand will get more attention paid to recursion and non-schema-based parsing, for completeness.

Ultimately, the fastest possible validation only api would be purpose built: it's much easier to tell if a number is valid than to parse it to a double.

**Comment by daverigby** [comment]
Thanks for the quick response John.

> It's accurate ... the fastest way to fully validate is probably to run it through simdjson's Dom api.

I'm less concerned about hitting say 4GB/s with the on-demand API (compared to quoted 2.5 GB/s with DOM API), more about the constant memory footprint. 

I didn't immediately see any memory figures on the website, but assuming the DOM API uses RAM roughly equal (or even 10% of) the input data size, that makes it difficult for me to use (can have many concurrent clients inserting new JSON documents, and having memory double (extra copy of DOM-ified document).

> It's possible to do with On Demand using recursion and a little elbow grease right now; I really don't know whether that will be faster or slower. In future releases on demand will get more attention paid to recursion and non-schema-based parsing, for completeness.

Thanks for confirming. I'll probably hang fire on trying to implement anything for now; certainly if there might be some expansions to the on-demand API in the not-too-distant-future.

**Comment by lemire** [comment]
> Ultimately, the fastest possible validation only api would be purpose built: it's much easier to tell if a number is valid than to parse it to a double.

That's where the ambiguity of the specification nails us. Is `231321321.321321321321e12332` a valid float in JSON? I would say no, but something people would disagree, and the specification allows us to disagree. There are definitively tricks that one can pull to try to validate numbers faster than you can parse them, but one would need to define the problem better.

**Comment by jkeiser** [comment]
Even at 1.0, on demand will probably still use an index, which can take up to 4x the document size in virtual memory (though generally takes up much, much less physical memory).
